File: FinalProject/app/src/main/java/com/example/finalproject/fragments/SpellCreationFragment.java
package com.example.finalproject.fragments;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Point;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.Display;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.example.finalproject.service.spell.Component;
import com.example.finalproject.service.spell.Element;
import com.example.finalproject.service.spell.Form;
import com.example.finalproject.service.spell.ManaChannel;
import com.example.finalproject.service.spell.ManaReservoir;
import com.example.finalproject.R;
import com.example.finalproject.service.spell.Spell;
import com.example.finalproject.service.spell.Type;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class SpellCreationFragment extends Fragment {
    private Element element=MainActivity.elements.get(0);
    private Type type=MainActivity.types.get(0);
    private Form form=MainActivity.forms.get(0);
    private ManaChannel mana_channel=MainActivity.manaChannels.get(0);
    private ManaReservoir mana_reservoir=MainActivity.manaReservoirs.get(0);
    private String name;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_spell_creation, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Display display = getActivity().getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        int width = size.x;
        final String[] spell_name = new String[1];
        EditText name_view=(EditText)getView().findViewById(R.id.spell_name);
        name_view.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                name=v.getText().toString();
                return false;
            }
        });
        Button confirm_spell=(Button)getView().findViewById(R.id.confirm_creation);
        name=name_view.getText().toString();
        confirm_spell.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                name=name_view.getText().toString();
                MainActivity.player.getSpells().add(new Spell(element, type, form, mana_channel, mana_reservoir, name));

            }
        });
        Button back=(Button)getView().findViewById(R.id.spell_creation_back_button);
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm=getParentFragmentManager();
                FragmentTransaction fragmentTransaction= fm.beginTransaction();
                fragmentTransaction.remove(fm.findFragmentById(R.id.spell_creation));
                fragmentTransaction.add(R.id.map, new MapFragment());
                fragmentTransaction.add(R.id.status, new StatusBarFragment());
                fragmentTransaction.add(R.id.menu, new MenuFragment());
                fragmentTransaction.commit();
            }
        });
        Bitmap bm=Bitmap.createBitmap(width/5, width/5, Bitmap.Config.ARGB_8888);
        bm.eraseColor(Color.YELLOW);
        Button element_view=(Button)getView().findViewById(R.id.element);
        Button mana_reservoir_view=(Button)getView().findViewById(R.id.mana_reservoir);
        Button type_view=(Button)getView().findViewById(R.id.type);
        Button mana_channel_view=(Button)getView().findViewById(R.id.mana_channel);
        Button form_view=(Button)getView().findViewById(R.id.form);
        RecyclerView comps=getView().findViewById(R.id.avaliable_components);
        comps.setLayoutManager(new LinearLayoutManager(getContext()));
        element_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                comps.setAdapter(new SpellAdapter(MainActivity.elements));
            }
        });
        mana_reservoir_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                comps.setAdapter(new SpellAdapter(MainActivity.manaReservoirs));
            }
        });
        mana_channel_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                comps.setAdapter(new SpellAdapter(MainActivity.manaChannels));
            }
        });
        type_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                comps.setAdapter(new SpellAdapter(MainActivity.types));
            }
        });
        form_view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                comps.setAdapter(new SpellAdapter(MainActivity.forms));
            }
        });
    }

    class SpellAdapter<T extends Component> extends RecyclerView.Adapter<SpellAdapter.ViewHolder>{
        private final ArrayList<T> data=new ArrayList<>();

        public SpellAdapter(ArrayList<T> data) {
            for (int i=0;i<data.size();i++){
                if (data.get(i).isAvailable())
                    this.data.add(data.get(i));
            }
        }

        public class ViewHolder extends RecyclerView.ViewHolder{
            TextView comp;
            public ViewHolder(@NonNull @NotNull View itemView) {
                super(itemView);
                comp=(TextView)itemView.findViewById(R.id.comp);
            }
        }

        @NonNull
        @NotNull
        @Override
        public SpellCreationFragment.SpellAdapter.ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
            View view=LayoutInflater.from(parent.getContext()).inflate(R.layout.components_item, parent, false);
            return new SpellAdapter.ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull @NotNull SpellCreationFragment.SpellAdapter.ViewHolder holder, int position) {
            Class r=data.get(position).getClass();
            holder.comp.setText(String.valueOf(data.get(position).getName()));
            if (r.equals(Element.class)) {
                holder.comp.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        FragmentManager fm=getChildFragmentManager();
                        FragmentTransaction fr=fm.beginTransaction();
                        element=(Element)data.get(position);
                        if (fm.findFragmentById(R.id.spell_characteristics)!=null)
                            fr.remove(fm.findFragmentById(R.id.spell_characteristics));
                        fr.add(R.id.spell_characteristics,
                                new SpellCharacteristicsFragment(
                                        new Spell(element,
                                                type,
                                                form,
                                                mana_channel,
                                                mana_reservoir,
                                                ((EditText)getView().findViewById(R.id.spell_name)).
                                                        getText().toString())));
                        fr.commit();
                    }
                });
            }
            else if (r.equals(Type.class)){
                holder.comp.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        FragmentManager fm=getChildFragmentManager();
                        FragmentTransaction fr=fm.beginTransaction();
                        type=(Type)data.get(position);
                        if (fm.findFragmentById(R.id.spell_characteristics)!=null)
                            fr.remove(fm.findFragmentById(R.id.spell_characteristics));
                        fr.add(R.id.spell_characteristics,
                                new SpellCharacteristicsFragment(
                                        new Spell(element,
                                                type,
                                                form,
                                                mana_channel,
                                                mana_reservoir,
                                                ((EditText)getView().findViewById(R.id.spell_name)).
                                                        getText().toString())));
                        fr.commit();
                    }
                });
            }
            else if (r.equals(ManaReservoir.class)){
                holder.comp.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        FragmentManager fm=getChildFragmentManager();
                        FragmentTransaction fr=fm.beginTransaction();
                        mana_reservoir=(ManaReservoir)data.get(position);
                        if (fm.findFragmentById(R.id.spell_characteristics)!=null)
                            fr.remove(fm.findFragmentById(R.id.spell_characteristics));
                        fr.add(R.id.spell_characteristics,
                                new SpellCharacteristicsFragment(
                                        new Spell(element,
                                                type,
                                                form,
                                                mana_channel,
                                                mana_reservoir,
                                                ((EditText)getView().findViewById(R.id.spell_name)).
                                                        getText().toString())));
                        fr.commit();
                    }
                });
            }
            else if (r.equals(ManaChannel.class)){
                holder.comp.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        FragmentManager fm=getChildFragmentManager();
                        FragmentTransaction fr=fm.beginTransaction();
                        mana_channel=(ManaChannel)data.get(position);
                        if (fm.findFragmentById(R.id.spell_characteristics)!=null)
                            fr.remove(fm.findFragmentById(R.id.spell_characteristics));
                        fr.add(R.id.spell_characteristics,
                                new SpellCharacteristicsFragment(
                                        new Spell(element,
                                                type,
                                                form,
                                                mana_channel,
                                                mana_reservoir,
                                                ((EditText)getView().findViewById(R.id.spell_name)).
                                                        getText().toString())));
                        fr.commit();
                    }
                });
            }
            else if (r.equals(Form.class)){
                holder.comp.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        FragmentManager fm=getChildFragmentManager();
                        FragmentTransaction fr=fm.beginTransaction();
                        form=(Form)data.get(position);
                        if (fm.findFragmentById(R.id.spell_characteristics)!=null)
                            fr.remove(fm.findFragmentById(R.id.spell_characteristics));
                        fr.add(R.id.spell_characteristics,
                                new SpellCharacteristicsFragment(
                                        new Spell(element,
                                                type,
                                                form,
                                                mana_channel,
                                                mana_reservoir,
                                                ((EditText)getView().findViewById(R.id.spell_name)).
                                                        getText().toString())));
                        fr.commit();
                    }
                });
            }
        }

        @Override
        public int getItemCount() {
            return data.size();
        }
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/fragments/SpellCreationFragment.kt
package com.example.finalproject.fragments

import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.finalproject.MainActivity
import com.example.finalproject.MainActivity.Companion.assets
import com.example.finalproject.R
import com.example.finalproject.service.classes.spell.Element
import com.example.finalproject.service.classes.spell.Form
import com.example.finalproject.service.classes.spell.ManaChannel
import com.example.finalproject.service.classes.spell.ManaReservoir
import com.example.finalproject.service.classes.spell.Spell
import com.example.finalproject.service.classes.spell.Type

class SpellCreationFragment : Fragment(), View.OnClickListener, TextView.OnEditorActionListener {

    private var element = assets.elements[0]
    private var type = assets.types[0]
    private var form = assets.forms[0]
    private var manaChannel = assets.manaChannels[0]
    private var manaReservoir = assets.manaReservoirs[0]
    private lateinit var nameView: EditText
    private lateinit var confirmSpell: Button
    private lateinit var back: Button
    private lateinit var elementView: Button
    private lateinit var manaReservoirView: Button
    private lateinit var typeView: Button
    private lateinit var manaChannelView: Button
    private lateinit var formView: Button
    private lateinit var comps: RecyclerView
    private var name: String = ""

    /**
     * inflates fragment's layout
     */
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_spell_creation, container, false)
    }

    /**
     * initializes graphic components
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        nameView = requireView().findViewById(R.id.spell_name)
        confirmSpell = requireView().findViewById(R.id.confirm_creation)
        back = requireView().findViewById(R.id.spell_creation_back_button)
        elementView = requireView().findViewById(R.id.element)
        manaReservoirView = requireView().findViewById(R.id.mana_reservoir)
        typeView = requireView().findViewById(R.id.type)
        manaChannelView = requireView().findViewById(R.id.mana_channel)
        formView = requireView().findViewById(R.id.form)
        comps = requireView().findViewById(R.id.avaliable_components)
        name = nameView.text.toString()
        nameView.setOnEditorActionListener(this)
        confirmSpell.setOnClickListener(this)
        back.setOnClickListener(this)
        comps.layoutManager = LinearLayoutManager(context)
        elementView.setOnClickListener(this)
        manaReservoirView.setOnClickListener(this)
        manaChannelView.setOnClickListener(this)
        typeView.setOnClickListener(this)
        formView.setOnClickListener(this)
    }
    private inner class SpellAdapter(dt: ArrayList<Int>) :
        RecyclerView.Adapter<SpellAdapter.ViewHolder>() {

        private val data: ArrayList<Int> = ArrayList()

        init {
            for (component in dt)
                if (assets.components[component]?.available == true)
                    data.add(component)
        }

        inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val comp: TextView = itemView.findViewById(R.id.comp)
        }

        /**
         * inflates the list item layout
         */
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.components_item, parent, false)
            return ViewHolder(view)
        }

        /**
         * @param holder a holder for a list item view
         * sets the text displayed in the list
         */
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.comp.text = assets.components[data[position]]?.name ?: ""
            holder.comp.setOnClickListener {
                when (data[position]) {
                    in assets.elements -> element = data[position]
                    in assets.types -> type = data[position]
                    in assets.forms -> form = data[position]
                    in assets.manaChannels -> manaChannel = data[position]
                    in assets.manaReservoirs -> manaReservoir = data[position]
                }
                val fm: FragmentManager = childFragmentManager
                val fr: FragmentTransaction = fm.beginTransaction()
                fm.findFragmentById(R.id.spell_characteristics)?.let { it1 -> fr.remove(it1) }
                fr.add(
                    R.id.spell_characteristics, SpellCharacteristicsFragment(
                        Spell(
                            assets.components[element] as Element,
                            assets.components[type] as Type,
                            assets.components[form] as Form,
                            assets.components[manaChannel] as ManaChannel,
                            assets.components[manaReservoir] as ManaReservoir,
                            (view!!.findViewById(R.id.spell_name) as EditText).text.toString()
                        )
                    )
                )
                fr.commit()
            }
        }

        /**
         * @return amount of items in the list
         */
        override fun getItemCount(): Int = data.size
    }

    /**
     * sets the click listener for needed views
     */
    override fun onClick(p0: View?) {
        when (p0) {
            confirmSpell -> {
                name = nameView.text.toString()
                MainActivity.player.spells.add(
                    Spell(
                        assets.components[element] as Element,
                        assets.components[type] as Type,
                        assets.components[form] as Form,
                        assets.components[manaChannel] as ManaChannel,
                        assets.components[manaReservoir] as ManaReservoir, name
                    )
                )
            }

            back -> {
                val fragmentManager = parentFragmentManager
                val fragmentTransaction = fragmentManager.beginTransaction()
                fragmentManager.findFragmentById(R.id.spell_creation)
                    ?.let { fragmentTransaction.remove(it) }
                fragmentTransaction.add(R.id.map, MapFragment(MainActivity.player.mapNumber))
                fragmentTransaction.add(R.id.status, StatusBarFragment())
                fragmentTransaction.add(R.id.menu, MenuFragment())
                fragmentTransaction.commit()
            }

            formView -> {
                comps.adapter = SpellAdapter(assets.forms)
            }

            typeView -> {
                comps.adapter = SpellAdapter(assets.types)
            }

            manaChannelView -> {
                comps.adapter = SpellAdapter(assets.manaChannels)
            }

            manaReservoirView -> {
                comps.adapter = SpellAdapter(assets.manaReservoirs)
            }

            elementView -> {
                comps.adapter = SpellAdapter(assets.elements)
            }
        }
    }

    /**
     * @param p0 the view that has been edited
     * sets the action on the event of editing a view
     */
    override fun onEditorAction(p0: TextView?, p1: Int, p2: KeyEvent?): Boolean {
        if (p0 == nameView) {
            name = p0.text.toString()
            return true
        }
        return false
    }
}